dfm/data/deduper.py

```diff
@@ -0,0 +1,224 @@
+        if self.split_method == "char_ngram":
+            max_char_idx = len(doc) - self.ngram_size
+            shingles = [
+                doc[i : i + self.ngram_size]
+                for i in range(0, max_char_idx, self.ngram_stride)
+            ]
+        elif self.split_method == "word_ngram":
+            words = doc.split(" ")
+            max_word_idx = len(words) - self.ngram_size
+            shingles = [
+                " ".join(words[i : i + self.ngram_size]).strip()
+                for i in range(0, max_word_idx, self.ngram_stride)
+            ]
+        elif self.split_method == "paragraph":
+            shingles = [p for p in doc.split("\n") if len(p) > 0]
+        elif self.split_method == "none":
+            shingles = [doc]
+        else:
+            raise ValueError(f"Invalid split method: {self.split_method}")
+
+        # Add all the shingles to the fingerprint
+        for shingle in shingles:
+            minhash.update(shingle.encode("utf-8"))
+
+        # Convert the fingerprint to a LeanMinHash fingerprint, to save memory
+        # and increase performance
+        minhash = LeanMinHash(minhash, seed=self.random_seed)
+
+        # Return the fingerprint
+        return minhash
+
+    def _store_document(
+        self, doc_idx: Union[str, int], doc: str, fname: Union[str, Path]
+    ):
+        """Appends the document to a JSONL file.
+
+        Args:
+            doc_idx (str or int): The document index.
+            doc (str): The document to append to the JSONL file.
+            fname (str or Path): The name of the JSONL file to append to.
+        """
+        # Ensure that `doc_id` is a string
+        doc_id = str(doc_idx)
+
+        # Ensure that `fname` is a Path object
+        fname = Path(fname)
+
+        # Append the document to the JSONL file
+        with fname.open("a") as f:
+            jsonned = json.dumps(dict(id=doc_id, text=doc))
+            f.write(jsonned + "\n")
+
+    def deduplicate(
+        self,
+        corpus: Union[Dataset, IterableDataset, Iterable[str]],
+        output_fname: Union[str, Path] = "deduplicated.jsonl",
+    ):
+        """Removes duplicate documents from the corpus and stores it to disk.
+
+        Args:
+            corpus (Dataset, IterableDataset or iterable of strings):
+                The corpus to deduplicate.
+            output_fname (str or Path, optional):
+                The name of the output file.
+
+        Raises:
+            FileExistsError: If the output file already exists.
+        """
+        # Convert corpus to an iterable of strings if a Dataset is given
+        if isinstance(corpus, Dataset) or isinstance(corpus, IterableDataset):
+            corpus = (sample["text"] for sample in corpus)
+
+        # Ensure that `output_fname` is a Path object
+        output_fname = Path(output_fname)
+
+        # If the output file already exists then raise an error
+        if output_fname.exists():
```

Comment:
You could consider including an `override` keyword argument, defaulting to `False`, instead of having to rename the output file.

Reply:
Good idea. Added now 😊